lib/utils/polymorphic.ts

```diff
@@ -0,0 +1,39 @@
+/** Little edits to original, source: https://www.benmvp.com/blog/polymorphic-react-components-typescript/ */
+import React from 'react';
+
+// Source: https://github.com/emotion-js/emotion/blob/master/packages/styled-base/types/helper.d.ts
+// A more precise version of just React.ComponentPropsWithoutRef on its own
+export type PropsOf<C extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<unknown>> =
+  JSX.LibraryManagedAttributes<C, React.ComponentPropsWithoutRef<C>>;
+
+interface AsProp<C extends React.ElementType> {
+  /**
+   * An override of the default HTML tag.
+   * Can also be another React component.
+   */
+  as?: C;
+}
+
+/**
+ * Allows for extending a set of props (`ExtendedProps`) by an overriding set of props
+ * (`OverrideProps`), ensuring that any duplicates are overridden by the overriding
+ * set of props.
+ */
+export type ExtendableProps<ExtendedProps = unknown, OverrideProps = unknown> = OverrideProps &
+  Omit<ExtendedProps, keyof OverrideProps>;
+
+/**
+ * Allows for inheriting the props from the specified element type so that
+ * props like children, className & style work, as well as element-specific
+ * attributes like aria roles. The component (`C`) must be passed in.
+ */
+export type InheritableElementProps<C extends React.ElementType, Props = unknown> = ExtendableProps<PropsOf<C>, Props>;
+
+/**
+ * A more sophisticated version of `InheritableElementProps` where
+ * the passed in `as` prop will determine which props can be included
+ */
+export type PolymorphicComponentProps<C extends React.ElementType, Props = unknown> = InheritableElementProps<
+  C,
+  Props & AsProp<C>
+>;
```

Comment:
Do we really need such a util helper and what is the license of this code?

Reply:
Maybe not needing, but helps to reuse polymorphic pattern in other components later down the road. Of course it can be merged with current component and later separated if some other component needs it too. Like normal text link.

License is not told in the article linked.